Fixed #31224 -- Added support for asynchronous views and middleware.

andrewgodwin opened the pull request:
ticket-31224 

This implements the ability to have asynchronous views in Django, by simply declaring a view as `async def`.

Under the hood, it rewrites the BaseHandler to be natively async and makes the WSGI request path run an async loop in-place for every request in order to service this. The ASGI request path down to a view is instead intended to be natively asynchronous all the way down.

This is still a work in progress, as it currently has several potential issues:

- [x] There are a small number of failing tests 
- [x] Synchronous middleware works, but its presence means a thread is used every request, defeating the point of async views. 
- [x] Class-based views do not have a clear path to async compatibility
- [x] There is not yet any documentation about how to use it

I'm putting it up as a draft PR so people have a place to easily see it and review the code so far.

auvipy commented:
is it too late to consider trio? for class-based async views, can we get inspiration from django-vanilla-views? what about making new async views instead of current class based views?

andrewgodwin commented:
@auvipy Unfortunately, while I like trio, it has a much smaller library ecosystem and so pragmatically we can't use it. My hope is eventually to make it possible to use in some situations.

Some more class-based views discussion is over on the forum (https://forum.djangoproject.com/c/internals/async) - while new async views is cleanest, it does mean having two versions of everything to maintain, which isn't great.

nediamond commented on tests/middleware_exceptions/middleware.py:
```diff
@@ -66,3 +67,21 @@ def process_template_response(self, request, response):
+        response.status_code = 418
+        return response
+
+    return middleware
```
Would it work to declare the async middleware like this?
```
class AsyncBaseMiddleware(BaseMiddleware):
    _is_async = True

class AsyncTeapotMiddleware(AsyncBaseMiddleware):
    async def __call__(self, request):
        response = await self.get_response(request)
        response.status_code = 418
        return response
```

charettes replied:
@nediamond I could be missing something but I don't Python offers a way to define an `async` `__call__` method.

In this case `AsyncTeapotMiddleware()()` returns a `coroutine` but `inspect.iscoroutinefunction(AsyncTeapotMiddleware())` returns `False`.

Maybe we could use `inspect.iscoroutinefunction(getattr(Foo, '__call__', None))` to avoid the `_is_async` annotation?

nediamond replied:
the following with an async `__call__` works for me!
```
class Test:
	async def __call__(self):
		return True
```
and
`inspect.iscoroutinefunction(Test.__call__) -> True`

charettes replied:
That's what `inspect.iscoroutinefunction(getattr(Foo, '__call__', None))` does above.

What I mean is that it's probably an abuse of Python data model. For example, 

```python
def Test:
    async def __iter__(self):
        pass

assert inspect.iscoroutinefunction(Test.__iter__)
```

Won't fail but `__aiter__` should be used for this purpose. There's no analogous `__acall__` for `__call__` and it's not clear to me whether `async def __call__` is an abuse or not.

nediamond replied:
Makes sense, looks like this is the relevant PEP section:
https://www.python.org/dev/peps/pep-0492/#why-magic-methods-start-with-a

not sure its an issue here since async `__call__` usage doesn't look to have keyword dependencies

andrewgodwin replied:
The initial comment is how async middleware works in this patch anyway - you `async def` the call, we detect it, and run it async. I'm not sure what change you're asking for?

nediamond replied:
Sorry, wasn't trying to request a change, was thinking about how async middleware would be written and just seeking clarification.

andrewgodwin replied:
Ah, no problem. The test middleware is maybe not the best example, I think once we convert a few of the built-in Django middlewares it'll be more obvious how you can write them.

charettes commented on django/core/handlers/base.py:
```diff
@@ -34,7 +37,12 @@ def load_middleware(self):
             middleware = import_string(middleware_path)
             try:
-                mw_instance = middleware(handler)
+                if getattr(middleware, "_is_async", False):
```
What about the following?

```suggestion
                if asyncio.iscoroutinefunction(getattr(middleware, '__call__', None)):
```

andrewgodwin commented:
@aeros We use the `iscoroutinefunction` function from `asyncio` as it allows us to flag things that are effectively coroutine functions (i.e. callables that return a coroutine) that are not declared as `async def`. Specifically, this is used in our SyncToAsync wrapper that turns sync-callables into async-callables.

By my reading, the `inspect` version does not offer this, whereas the `asyncio` version lets you annotate functions with `asyncio.coroutines._is_coroutine` to mark them as such. Given this, we can't use the `inspect` version, as it's too narrow - it only returns functions that were declared async def, rather than actually telling you if something is a coroutine function.

aeros commented:
> By my reading, the inspect version does not offer this, whereas the asyncio version lets you annotate functions with asyncio.coroutines._is_coroutine to mark them as such. Given this, we can't use the inspect version, as it's too narrow - it only returns functions that were declared async def, rather than actually telling you if something is a coroutine function.

@andrewgodwin That's correct. I mainly wanted to make sure that it was intentional and it's clear that doing so will indirectly support the older generator coroutines that are being deprecated in Python 3.10 (unless `asyncio.iscoroutine()` and `asyncio.iscoroutinefunction()` are changed). I wasn't aware of Django's `SyncToAsync` wrapper, but I can see why `inspect.iscoroutine()` would not be compatible with it.

weargoggles commented on django/core/handlers/base.py:
```diff
@@ -82,6 +144,28 @@ def get_response(self, request):
             )
         return response
 
+    async def get_response_async(self, request):
```
Might this be more at home in ASGIHandler?

andrewgodwin replied:
I didn't want to put it there in case someone wanted to implement an async-capable handler that was not based on ASGI. It's the same reasoning `get_response` is here, not in WSGIHandler.

andrewgodwin commented:
Given that this branch now passes all tests and shows no drop in sync performance in my testing, I consider it ready for review. I've also added relevant docs.

claudep commented on django/utils/deprecation.py:
```diff
@@ -80,6 +80,10 @@ def __instancecheck__(self, instance):
+
+    sync_capable = True
+    async_capable = False
+
```
I think we need to have a discussion about the `MiddlewareMixin` class. Currently, it's mostly advertised as a deprecation tool to convert to new-style middleware. However, I tried recently to get rid of it in DJango, and we are in fact blocked by the middleware decorating facility, as if the pre- and the -post `get_response` code is embedded in the same `__call__` method, it's very difficult to implement a decoration tool for that. @apollo13, I would be interested to know if you had a plan for that use case.

So now, if we recognize that `MiddlewareMixin` is useful not only for the deprecation part, but also for decoration and here for sync/async properties, we should move it to somewhere in `django/middleware` and update the custom middleware documentation to suggest inheriting from that class.

apollo13 replied:
@claudep Is "no idea, I hoped a clear answer" would emerge sooner or later enough? Joking aside, I do not have good answers :(

claudep replied:
:-)
So you wouldn't be shocked to see us keeping some `django.middleware.MiddlewareMixin` to keep the decoratable functionality (without changing the basic middleware contract)?

apollo13 replied:
I wouldn't be shocked, but it really goes against "there should be one way of doing things". So all in all if anyone has ideas to finish what we started with the middleware move I'd happily use that over the status quo.

Is the code of your efforts online somewhere? Maybe we find a way to fix decorators :)

claudep replied:
Part of my work is https://github.com/django/django/pull/11828 (I'll rebase soon). But then I blocked on deprecating `MiddlewareMixin` due to decoration.

andrewgodwin replied:
Yes, I was sort of shocked when I found that basically all core middleware still uses `MiddlewareMixin`. I think there's still a path away from it, though - but maybe to more callable-style classes, rather than pure factory functions.

nicolaslara commented on django/core/handlers/base.py:
```diff
@@ -30,11 +35,38 @@ def load_middleware(self):
+                    mw_instance = middleware(handler)
+                elif middleware_is_async and not handler_is_async:
+                    if settings.DEBUG:
+                        logger.debug('Synchronous middleware adapted: %r', middleware_path)
```
This wouldn't happen on every request, just on server restart (when `load_middleware()` is called)

jacobian replied:
A system check seems a good option if the warning would be too noisy.

carltongibson commented on django/core/handlers/wsgi.py:
```diff
@@ -130,7 +133,13 @@ def __call__(self, environ, start_response):
+
+        # Get the response, either using async get_response or standard.
+        if asyncio.iscoroutinefunction(self.get_response):
+            response = async_to_sync(self.get_response)(request)
```
Is this block correct? 

`self.get_response()` is `BaseHandler.get_response()`, which in contrast to `get_response_async()` is **not** `async def`, so will this branch ever evaluate as `True`? 

The contrasting point in ASGIHandler just does `await self.get_response_async(request)`, allowing the branching inside `_get_response_async()` once the callback is resolved. 

This line isn't hit by the test suite.

andrewgodwin replied:
This is a relic of the previous attempt - looks like I got the one in `ASGIHandler` but not this. Pushed a commit to remove it.

andrewgodwin commented:
Good point about the testing - I always tend to forget that late in a project (especially as I was mostly performance-testing this with a separate suite). I'll try to get some extra tests in today or over the weekend and get that coverage up.

andrewgodwin commented:
@carltongibson I've added more handler tests as well as the necessary `AsyncClient` we needed to support them, and then some user docs about using all this for your own async tests.

carltongibson commented:
Super @andrewgodwin! I saw the commit come in and had to laugh. Great work. Will look again tomorrow.

andrewgodwin commented:
@carltongibson Ah, interesting! I only ever run the test suites rather than the individual methods, so I never ran into this.

Let me go dive in and see if I can figure out a way to have Django do this automatically; I had considered making the test runner just auto-apply `async_to_sync`, so maybe this is the time to do this.

carltongibson commented:
Hey @andrewgodwin Super. Thanks.

Whilst you're looking... My other (currently proto-) question is, Can we reduce the number of `async def`s in test/client.py? 

* Does AsyncRequestFactory need `async def` at all? 
   * If yes, can it just be `request()`, but that doesn't await anything so...?
* Then we just have `async def request()` in AsyncClient, which does `await`. 

Am still playing with it, but thought I'd mention it early since you're here. 🙂


Other than that, this is looking good I think. (Time to ask @felixxm to get out the fine-toothed comb, before another fine read of the docs...) Thanks for the super effort.

andrewgodwin commented:
@carltongibson I don't think I can stop `AsyncRequestFactory.request` being one - since its child `AsyncClient.request` is - but I might be able to remove it from `AsyncRequestFactory.generic`. Will have a go.

carltongibson commented:
I think that's allowed: 

```
import asyncio


class Parent:
    def foo(self):
        return "Hello"


class Child(Parent):
    async def foo(self):
        return super().foo()


async def main():
    c = Child()
    return await c.foo()


if __name__ == '__main__':
    print(
        asyncio.run(main())
    )
```

Runs as 

```
~/Desktop $ python async_subclass.py 
Hello
```

Nothing seemed to break when I was experimenting earlier. 


---

I see you rebased. Failures are from 4e8d6a1bafbbebc0720c5d1ed751521356364fe9

andrewgodwin commented:
@carltongibson It is possible, definitely, but I don't think it is a good idea - IMO, the types signatures of parent and child should always match. I did get rid of async def on `generic`, though.

I'm just in the process of trying to make the middleware a wee bit more efficient by making `MiddlewareMixin` async-aware and then I'll properly rebase and fix the conflicts/tests.

carltongibson commented:
@andrewgodwin OK, that's a different kettle of fish. Happy if that's your reasoning. 

Super stuff. Thanks 👍

carltongibson commented:
Hey @andrewgodwin 

> ... the types signatures of parent and child should always match ...

But we already swapped from def to async def when we subclassed RequestFactory for AsyncRequestFactory. (So we could just delay that one no?)

(More than the particular answer, I´m wondering if there isn´t a little _Smell_ here. I´m sure it´d be _Fine_, but I´d like us to get it right, because once it´s in I think we´re stuck with it.)

andrewgodwin commented:
> But we already swapped from def to async def when we subclassed RequestFactory for AsyncRequestFactory. (So we could just delay that one no?)

Hm, you got me there. Personally, I would have made every method on `AsyncRequestFactory` explicitly async, but refrained from doing so to avoid useless code (they are all still async by convention, as they return coroutines, but they will not return `True` if you look at them with `iscoroutinefunction`, so they're in a weird middle ground).

I'll remove it for now, and put a comment to advise the casual source code reader what is happening.

aeros commented:
@carltongibson While that's technically allowed, if that await never eventually hits a `yield` (in an `__await__`), the coroutine will never get suspended. This means means the await ends up being fairly pointless, and makes `await Child().foo()` behave effectively the same as if you had just called `Parent().foo()` directly.

andrewgodwin commented:
Alright, I've rebased to master, squashed, fixed up the tests to remove the need to even have `async_to_sync` except in rare cases, as well as made all the default middleware considerably more efficient in async mode (no longer uses a thread while the view is running) and added a 3.1 release notes section.

I await further feedback and the fine-toothed comb!

andrewgodwin commented on docs/spelling_wordlist:
```diff
@@ -25,6 +25,7 @@ arg
 args
 assistive
 async
+asynchronous
```
It wasn't acceptable to the CI job last time it ran; I'll remove it again and try!

andrewgodwin commented on django/utils/deprecation.py:
```diff
@@ -80,15 +83,46 @@ def __instancecheck__(self, instance):
+        if asyncio.iscoroutinefunction(self.get_response):
+            # We mark the class as async-capable, but the actual switch
+            # is done inside __call__ to avoid swapping out dunder methods
+            self._is_coroutine = asyncio.coroutines._is_coroutine
```
There is no way to do this officially, and it's one of the several things that annoys me about `asyncio`. Twisted, however, also relies on this particular detail (I believe they asked for it to be added), so between them and us, I imagine we can keep it in - at least until Python grows a better way of doing this.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -109,43 +184,144 @@ def _get_response(self, request):
-                "returned None instead." % (callback.__module__, view_name)
-            )
+        # Complain if the view returned None or an uncalled coroutine
+        self.check_none_response(response, callback)
```
I moved `check_none_response()` to a separate PR #12474 and renamed it to the `check_response()`. I'm going to rebase this patch after merging.

andrewgodwin replied:
Works for me - I don't mind either way.

felixxm commented:
I merged autoconversion of test async methods in 17009e910538b9c3709d302f54c2d10eb245cf38 and rebased this PR.

andrewgodwin commented:
@felixxm Looks good to me. Any more work needed on this PR, or are we just waiting for final review and merge?

felixxm commented:
@andrewgodwin We're waiting for me, sorry. I'm working on small edits and final review. I'm not an async expert so I want to create few sample projects and understand all of this :face_with_head_bandage:.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -137,20 +215,83 @@ def _get_response(self, request):
+
+        if response is None:
+            wrapped_callback = self.make_view_atomic(callback)
+            # If it is a synchronous view, run it in a subthread
```
Is this comment correct?

andrewgodwin replied:
Yes, it is correct.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -137,20 +215,83 @@ def _get_response(self, request):
+        if hasattr(response, 'render') and callable(response.render):
+            for middleware_method in self._template_response_middleware:
+                response = await middleware_method(request, response)
+                # Complain if the template response middleware returned None (a common error).
```
`... or an uncalled coroutine.`?

felixxm commented on django/core/handlers/base.py:
```diff
@@ -137,20 +215,83 @@ def _get_response(self, request):
+                    name='%s.process_template_response' % (
+                        middleware_method.__self__.__class__.__name__,
+                    )
+                )
```
This is untested.

andrewgodwin replied:
It was already untested previously, unfortunately. Do we have any middleware that supports deferred rendering we can slot into a test easily?

felixxm replied:
Similar branch in `_get_response()` is covered by `tests.middleware_exceptions`. I would try to add tests for it.

felixxm replied:
I added tests for all paths in `_get_async_response()`.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -137,20 +215,83 @@ def _get_response(self, request):
+
+        # Make sure the response is not a coroutine
+        if asyncio.iscoroutine(response):
+            raise RuntimeError("Response is still a coroutine")
```
Is this possible? If so, it will be good to cover this scenario with tests.

andrewgodwin replied:
It is not possible currently, but it's a very easy mistake to make when working on this code and so I thought it would be good to leave it in there as a protection for anyone fiddling with it in future.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -82,24 +133,34 @@ def get_response(self, request):
+                '%s: %s', response.reason_phrase, request.path,
+                response=response,
+                request=request,
+            )
```
I added test for this message.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -160,10 +287,41 @@ def check_response(self, response, callback, name=None):
+                if asyncio.iscoroutinefunction(view):
+                    raise RuntimeError(
+                        'You cannot use ATOMIC_REQUESTS with async views.'
+                    )
```
I added test for this error.

felixxm commented on django/core/handlers/base.py:
```diff
@@ -137,20 +201,83 @@ def _get_response(self, request):
+            try:
+                response = await wrapped_callback(request, *callback_args, **callback_kwargs)
+            except Exception as e:
+                response = await sync_to_async(self.process_exception_by_middleware, thread_sensitive=True)(e, request)
```
I'm not sure if wrapping with `sync_to_async()` is required :thinking: , because it's already adapted in [BaseHandler.load_middleware()](https://github.com/django/django/pull/11650/files#diff-dbd7d9159676b15fc9a096b0adb919e9R82-R84). I don't see any failure after changing to:
```python
response = await self.process_exception_by_middleware(e, request)
```

andrewgodwin replied:
You can't just `await` a non-async function like that, unfortunately, it won't work. The exception methods were adapted, but the wrapper was not; I've instead changed it for now so `process_exception` is always run synchronously, to avoid duplicating this request path (and since it does not need to be super performance-sensitive anyway).

carltongibson commented on docs/topics/async.txt:
```diff
@@ -4,18 +4,110 @@ Asynchronous support
+    Django uses ``asyncio.iscoroutinefunction`` to test if your view is
+    asynchronous or not. If you implement your own method of returning a
+    coroutine, ensure you set the ``_is_coroutine`` attribute of the view
+    to ``asyncio.coroutines._is_coroutine`` so this function returns ``True``.
```
(Ref @felixxm's https://github.com/django/django/pull/11650/files#r380653311, which is fine (+1).)

Are we sure we want to **document** this in Django's docs? Maybe 3.9/3.10 arrive and there's an official way to do this, and we can just point to that. 

("But we support older versions" — sure: inclined to say, "if you're using the async stuff, use the latest Python, because asyncio is still evolving rapidly"...) 

Not sure. Maybe it's OK, but what do we think?

andrewgodwin replied:
I want to call it out because it is an entirely non-obvious behaviour that the Python docs repeatedly gloss over. If I kept forgetting about it and needed reminding, I want our advanced users to have this told to them directly.

carltongibson replied:
OK. Makes sense.

carltongibson commented on docs/topics/async.txt:
```diff
@@ -4,18 +4,110 @@ Asynchronous support
+dealing with external APIs or datastores.
+
+If you want to call a part of Django that is still synchronous (like the ORM)
+you will need to wrap it in a ``sync_to_async`` call, like this::
```
I feel there's a link to the `asgiref` docs missing somewhere — it's not really discussed. 

(Can we do Sphinx links for those...? 🤔)

andrewgodwin replied:
I've omitted linking to the asgiref docs purely because there's not much there of substance; async_to_sync and sync_to_async are documented more in the Django codebase than in asgiref (which is more general-ASGI docs).

carltongibson replied:
I have a lingering feeling they need some documentation. Maybe just a section here, saying what they are?

andrewgodwin replied:
Yup, I'll write up some reference docs quickly and put them at the bottom there for now.

andrewgodwin replied:
OK, docs added. They're a little long but I feel that's better than being too short.

carltongibson replied:
Super. Thanks @andrewgodwin. I'll give it a look tomorrow now. Have a good rest of your day!

andrewgodwin commented:
I've patched a couple of the things you called out, @felixxm - I think the rest are addressed in comments. Let me know if not.

@carltongibson your doc patch looks perfectly fine to me! Please squash it in along with my most recent little commit.

felixxm commented:
I added `AsyncMiddlewareTests.test_process_template_response_returns_none_with_sync_client()` that fails with `AttributeError: 'AsyncToSync' object has no attribute '__self__'` :thinking:

felixxm commented:
@andrewgodwin https://github.com/django/asgiref/pull/150 fixes this test for me.

andrewgodwin commented:
Thanks - there's one niggle there with `==` versus `is` and then we can merge and release that to get this passing.

felixxm commented:
I squashed by tests with the main commit and rebased from master.

carltongibson commented:
I squashed in my previous edits, and fixed the build for the wrapper functions docs. (I'll comment on those separately.)

carltongibson commented:
Hey @andrewgodwin. 😅 I feel like we're getting there. I've added a fixup commit with suggested edits for the wrapper functions docs. (Small rewording, plus sphinx links to the two functions from everywhere.) 

Please review at your leisure, and we can squash down if you're happy.

felixxm commented:
@andrewgodwin In the last commit aff07dc527f9ae6b91c68ed741258fbd15d74045 I tried to use `AsyncRequestFactory` in `asgi.tests`. Can you take a look?

andrewgodwin commented:
@carltongibson That fixup looks good to me - thanks for doing all the Sphinx linking that I often forget!

@felixxm That looks like a sensible change to me, nice way of reducing the scope duplication.

felixxm commented:
I moved documentation of async adapter functions to a separate PR #12562.

felixxm commented:
@andrewgodwin I think `AsyncRequestFactory` is missing in `docs/topics/testing/advanced.txt` :thinking: (`django/test/client.py` is the last file that I'm checking/polishing, so we're really close :1st_place_medal:).

felixxm commented:
I push small changes:
- made ``AsyncClient`` and ``AsyncRequestFactory`` importable from ``django.test``,
- added basic tests for HTTP methods of ``AsyncRequestFactory``.

felixxm commented:
buildbot, test on selenium.

andrewgodwin commented:
I added a small section about ``AsyncRequestFactory`` in the docs just now, too.

felixxm commented on docs/topics/testing/tools.txt:
```diff
@@ -1755,6 +1755,60 @@ You can also exclude tests by tag. To run core tests if they are not slow:
+asynchronous test client. This is available as ``django.test.AsyncClient``,
+or as ``self.async_client`` on any test.
+
+It has the same methods and signatures as the synchronous (normal) test client,
```
This is not entirely true, we should at least add a note that HTTP methods: `get()`, `post()`, `head()`, `options()`, `put()`, `patch()`, and `trace()` doesn't support the `follow` argument.

carltongibson replied:
I adjusted `AsyncClient` to raise `NotImplemented` is you pass the `follow` parameter, plus tests in d678b36c1c831b5b54b007cc1bf47be644b8e5a2.

felixxm commented on django/test/client.py:
```diff
@@ -714,3 +848,62 @@ def _handle_redirects(self, response, data='', content_type='', **extra):
+        response.context = data.get('context')
+        response.json = partial(self._parse_json, response)
+        # Attach the ResolverMatch instance to the response.
+        response.resolver_match = SimpleLazyObject(lambda: resolve(request['PATH_INFO']))
```
I think this should use the `request['path']`.

felixxm replied:
I fixed this issue in c86f88bc6f8fdf1f76c51462c944e0eb223c867e.

felixxm commented:
@andrewgodwin Can you take a look at our recent changes?

felixxm commented:
buildbot, test on selenium.

felixxm commented:
buildbot, test on oracle.

carltongibson commented:
I just pushed 40ca60c10f09edc0a24acf756c8daf471f8b82de, which adds an:

```
await communicator.await() 
```

... at the end of the FileResponse test case `asgi.tests.ASGITest.test_file_response`

Previously we had a:

```
Exception ignored in: <_io.FileIO name='/Users/carlton/Documents/Django-Stack/django/tests/asgi/urls.py' mode='rb' closefd=True>
Traceback (most recent call last):
  File "/Users/carlton/.pyenv/versions/3.9.0a4/lib/python3.9/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
ResourceWarning: unclosed file <_io.BufferedReader name='/Users/carlton/Documents/Django-Stack/django/tests/asgi/urls.py'>
```

This was caused by the `sync_to_async(response.close)` future getting a `asyncio.exceptions.CancelledError` when the `communcator` went out of scope at the end of the test case. 

We just need to make sure the app has a change to finish in these cases. 
(Just noting for future. 🙂)

andrewgodwin commented:
Yup, all these edits look good to me. Thanks to both of you!

carltongibson commented:
Cheers @andrewgodwin. Thanks for the confirmation. 

I think we're there 🤞. @felixxm can press the button. 😀

💃⛵️🎇

felixxm commented:
Thanks y'all :1st_place_medal: :rocket: :confetti_ball:

andrewgodwin commented:
Thanks so much for all your hard work reviewing!

Now, let's take a look at that ORM... :)

jezdez commented:
@andrewgodwin 🤯!

stlk commented:
Thank you for this!
